File: Vachanam/Document/Parsers/PlainTextParser.swift
//
//  PlainTextParser.swift
//  Vachanam
//
//  Parses raw plain text and TXT files into structured ParsedDocument models.
//

import Foundation

public struct PlainTextParser: DocumentParser {
    public init() {}
    
    public static func decodeString(from data: Data) -> String {
        if let str = String(data: data, encoding: .utf8) { return str }
        if let str = String(data: data, encoding: .isoLatin1) { return str }
        if let str = String(data: data, encoding: .windowsCP1252) { return str }
        if let str = String(data: data, encoding: .utf16) { return str }
        return String(data: data, encoding: .ascii) ?? ""
    }
    
    public func parse(from source: DocumentSource) async throws -> ParsedDocument {
        let text: String
        let defaultTitle: String
        
        switch source {
        case .fileURL(let url):
            guard FileManager.default.fileExists(atPath: url.path) else {
                throw DocumentParserError.fileNotFound(url)
            }
            guard let data = try? Data(contentsOf: url) else {
                throw DocumentParserError.parsingFailed("Could not read plain text file: \(url.lastPathComponent)")
            }
            text = Self.decodeString(from: data)
            defaultTitle = url.deletingPathExtension().lastPathComponent
        case .rawText(let rawText, let rawTitle):
            text = rawText
            defaultTitle = rawTitle
        case .webURL:
            throw DocumentParserError.invalidSource("PlainTextParser requires a fileURL or rawText")
        }
        
        let normalized = text.replacingOccurrences(of: "\r\n", with: "\n").replacingOccurrences(of: "\r", with: "\n")
        var rawParagraphs = normalized.components(separatedBy: "\n\n")
            .map { $0.replacingOccurrences(of: "\n", with: " ").trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
        
        // Fallback for files with single-newline separation
        if rawParagraphs.count <= 1 && normalized.contains("\n") {
            let lines = normalized.components(separatedBy: "\n")
                .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
                .filter { !$0.isEmpty }
            if lines.count > 1 {
                rawParagraphs = lines
            }
        }
        
        let trimmed = normalized.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !rawParagraphs.isEmpty || !trimmed.isEmpty else {
            throw DocumentParserError.parsingFailed("Document is empty")
        }
        if rawParagraphs.isEmpty {
            rawParagraphs = [trimmed]
        }
        
        var title = defaultTitle
        var startIndex = 0
        
        // If the first paragraph is short, treat it as the document title
        if let first = rawParagraphs.first, first.count <= 80 && !first.hasSuffix(".") {
            title = first
            startIndex = 1
        }
        
        let contentParagraphs = startIndex < rawParagraphs.count ? Array(rawParagraphs[startIndex...]) : rawParagraphs
        let blocks = contentParagraphs.map { ParsedBlock(type: .paragraph, text: $0) }
        
        let chapter = ParsedChapter(title: title, blocks: blocks)
        return ParsedDocument(title: title, author: nil, format: .plainText, chapters: [chapter])
    }
}

File: VachanamAndroid/app/src/main/java/com/vachanam/reader/data/parser/PlainTextParser.kt
package com.vachanam.reader.data.parser

import com.vachanam.reader.data.model.BlockType
import com.vachanam.reader.data.model.DocumentFormat
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileNotFoundException
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets

class PlainTextParser : DocumentParser {

    companion object {
        fun decodeBytes(bytes: ByteArray): String {
            val encodings = listOf(
                StandardCharsets.UTF_8,
                StandardCharsets.ISO_8859_1,
                Charset.forName("windows-1252"),
                StandardCharsets.UTF_16,
                StandardCharsets.US_ASCII
            )
            for (charset in encodings) {
                try {
                    val decoded = String(bytes, charset)
                    if (decoded.isNotEmpty()) return decoded
                } catch (_: Exception) {}
            }
            return String(bytes, StandardCharsets.UTF_8)
        }
    }

    override suspend fun parse(source: DocumentSource): ParsedDocument = withContext(Dispatchers.IO) {
        val (text, defaultTitle) = when (source) {
            is DocumentSource.LocalFile -> {
                if (!source.file.exists()) throw FileNotFoundException("File not found: ${source.file.absolutePath}")
                val bytes = source.file.readBytes()
                Pair(decodeBytes(bytes), source.file.nameWithoutExtension)
            }
            is DocumentSource.RawText -> {
                Pair(source.text, source.title)
            }
            is DocumentSource.WebUrl -> {
                throw IllegalArgumentException("PlainTextParser cannot parse WebUrl directly")
            }
        }

        val normalized = text.replace("\r\n", "\n").replace("\r", "\n")
        var rawParagraphs = normalized.split("\n\n")
            .map { it.replace("\n", " ").trim() }
            .filter { it.isNotEmpty() }

        if (rawParagraphs.size <= 1 && normalized.contains("\n")) {
            val lines = normalized.split("\n").map { it.trim() }.filter { it.isNotEmpty() }
            if (lines.size > 1) {
                rawParagraphs = lines
            }
        }

        if (rawParagraphs.isEmpty()) {
            val trimmed = normalized.trim()
            if (trimmed.isEmpty()) throw IllegalStateException("Document is empty")
            rawParagraphs = listOf(trimmed)
        }

        var title = defaultTitle
        var startIndex = 0

        // If the first paragraph is short, treat it as the document title
        val first = rawParagraphs.firstOrNull()
        if (first != null && first.length <= 80 && !first.endsWith(".")) {
            title = first
            startIndex = 1
        }

        val contentParagraphs = if (startIndex < rawParagraphs.size) {
            rawParagraphs.subList(startIndex, rawParagraphs.size)
        } else {
            rawParagraphs
        }

        val blocks = contentParagraphs.map {
            ParsedBlock(
                type = BlockType.PARAGRAPH,
                text = it
            )
        }

        val chapter = ParsedChapter(title = title, blocks = blocks)
        ParsedDocument(
            title = title,
            author = null,
            format = DocumentFormat.PLAIN_TEXT,
            chapters = listOf(chapter)
        )
    }
}
